beacon_node/store/src/hot_cold_store.rs

```diff
@@ -0,0 +1,222 @@
+use crate::chunked_vector::{
+    store_updated_vector, ActiveIndexRoots, BlockRoots, CompactCommitteesRoots, HistoricalRoots,
+    RandaoMixes, StateRoots,
+};
+use crate::iter::{ReverseStateRootIterator, StateRootsIterator};
+use crate::{leveldb_store::LevelDB, DBColumn, Error, PartialBeaconState, Store, StoreItem};
+use parking_lot::RwLock;
+use slog::{info, trace, Logger};
+use std::convert::TryInto;
+use std::path::Path;
+use std::sync::Arc;
+use types::*;
+
+pub struct HotColdDB {
+    /// The slot before which all data is stored in the cold database.
+    ///
+    /// Data for slots less than `split_slot` is in the cold DB, while data for slots
+    /// greater than or equal is in the hot DB.
+    split_slot: Slot,
+    /// Cold database containing compact historical data.
+    cold_db: LevelDB,
+    /// Hot database containing duplicated but quick-to-access recent data.
+    hot_db: LevelDB,
+    /// Chain spec.
+    spec: ChainSpec,
+    /// Logger.
+    pub(crate) log: Logger,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum HotColdDbError {
+    FreezeSlotError {
+        current_split_slot: Slot,
+        proposed_split_slot: Slot,
+    },
+}
+
+impl Store for HotColdDB {
+    // Defer to the hot database for basic operations (including blocks for now)
+    fn get_bytes(&self, column: &str, key: &[u8]) -> Result<Option<Vec<u8>>, Error> {
+        self.hot_db.get_bytes(column, key)
+    }
+
+    fn put_bytes(&self, column: &str, key: &[u8], value: &[u8]) -> Result<(), Error> {
+        self.hot_db.put_bytes(column, key, value)
+    }
+
+    fn key_exists(&self, column: &str, key: &[u8]) -> Result<bool, Error> {
+        self.hot_db.key_exists(column, key)
+    }
+
+    fn key_delete(&self, column: &str, key: &[u8]) -> Result<(), Error> {
+        self.hot_db.key_delete(column, key)
+    }
+
+    /// Store a state in the store.
+    fn put_state<E: EthSpec>(
+        &self,
+        state_root: &Hash256,
+        state: &BeaconState<E>,
+    ) -> Result<(), Error> {
+        if state.slot <= self.split_slot {
+            self.store_archive_state(state_root, state)
+        } else {
+            self.hot_db.put_state(state_root, state)
+        }
+    }
+
+    /// Fetch a state from the store.
+    fn get_state<E: EthSpec>(
```

Comment:
WRT needing additional additional fields in put/get, I was thinking we could split `StoreItem` into get/put and do :

```rust
impl GetStoreItem for (BeaconState<E>, Option<Slot>) { ... }

impl PutStoreItem for BeaconState<E> { ... }
```

However, I'm not sure that's better than what you have here. I'm just mentioning it FYI.

Reply:
Why did you change your mind on this? I was thinking it might be quite nice, and would resolve your concern about phase1 as a first class citizen